fuzzer/src/buttercup/fuzzing_infra/runner_proxy.py

```diff
@@ -0,0 +1,215 @@
+import json
+import logging
+import os
+import signal
+import subprocess
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Any
+
+from buttercup.common.types import FuzzConfiguration
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Crash:
+    input_path: str
+    stacktrace: str
+    reproduce_args: list[str]
+    crash_time: float
+
+
+@dataclass
+class FuzzResult:
+    """Result from a fuzzing operation"""
+
+    logs: str
+    command: str
+    crashes: list[Crash]
+    stats: dict
+    time_executed: float
+    timed_out: bool
+
+
+@dataclass
+class Conf:
+    # in seconds
+    timeout: int
+    runner_path: Path
+
+
+@dataclass
+class RunnerProxy:
+    conf: Conf
+    _timeout: int = field(init=False, default=5)
+
+    def _kill_process(self, process: subprocess.Popen) -> None:
+        try:
+            process.kill()
+            try:
+                process.wait(timeout=self._timeout)
+            except subprocess.TimeoutExpired:
+                logger.error("Process did not terminate after kill within 5 seconds")
+        except Exception as e:
+            logger.error(f"Error killing process: {e}")
+
+    def _kill_process_group(self, process: subprocess.Popen) -> None:
+        # Kill the entire process group
+        try:
+            if os.name != "nt":  # Unix-like systems
+                # Kill the process group
+                os.killpg(os.getpgid(process.pid), signal.SIGTERM)
+                # Wait a bit for graceful termination
+                try:
+                    process.wait(timeout=self._timeout)
+                except subprocess.TimeoutExpired:
+                    # Force kill if graceful termination failed
+                    os.killpg(os.getpgid(process.pid), signal.SIGKILL)
+                    try:
+                        process.wait(timeout=self._timeout)
+                    except subprocess.TimeoutExpired:
+                        logger.error("Process did not terminate after kill within 5 seconds")
+            else:  # Windows
+                self._kill_process(process)
+        except (ProcessLookupError, OSError) as e:
+            logger.warning(f"Error killing process group: {e}")
+            self._kill_process(process)
+
+    def _run_subprocess_task(self, cmd: list[str], timeout: int, task_type: str) -> dict[str, Any]:
+        """Run subprocess task and return result"""
+        # Enforce a timeout of runner_timeout + 5 minutes (in seconds)
+        subprocess_timeout = timeout + 300
+
+        try:
+            process = subprocess.Popen(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                start_new_session=True,
+            )
+
+            try:
+                stdout, stderr = process.communicate(timeout=subprocess_timeout)
```

Comment:
Is there a risk that either stdout or stderr becomes huge here? Should we consider a limit on the amount of data read? IIRC `.communicate()` is how you are supposed to do things, but just thinking that we might want to cut in the beginning or middle of the output if it becomes huge and that would probably require explicit reading of the fds. Maybe create an issue and ignore for now?
Maybe such an issue would manifest at earlier stages first, in the runner?

Reply:
Yes, good point. I can fix it quickly I think.